File: 数组/34. 在排序数组中查找元素的第一个和最后一个位置/solution.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = -1, r = nums.size();
        int i, j;
        if(nums.empty()) return vector<int>{-1, -1};
        if(nums.size() == 1) {
            if(nums[0] == target) return vector<int>{0, 0};
            else return vector<int>{-1, -1};
        }
        while(l + 1 < r) {
            int mid = l + (r - l) / 2;
            if(nums[mid] >= target) r = mid;
            else l = mid;
            i = r;
        }
        if(r == nums.size() || nums[i] != target) return vector<int>{-1, -1};
        else {
            int l = -1, r = nums.size();
            while(l + 1 < r) {
                int mid = l + (r - l) / 2;
                if(nums[mid] <= target) l = mid;
                else r = mid;
                j = l;
            }
        }
        return vector<int>{i, j};
    }
};

File: 数组/34. 在排序数组中查找元素的第一个和最后一个位置/solution.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        l, r = -1, n
        if not nums:
            return [-1, -1]
        if n == 1:
            if nums[0] == target:
                return [0, 0]
            else:
                return [-1, -1]
        while l + 1 < r:
            mid = l + (r - l) // 2
            if nums[mid] >= target:
                r = mid
            else:
                l = mid
            i = r
        if r == n or nums[i] != target:
            return [-1, -1]
        else:
            l, r = -1, n
            while l + 1 < r:
                mid = l + (r - l) // 2
                if nums[mid] <= target:
                    l = mid
                else:
                    r = mid
                j = l
        return [i, j]